modules/h2/src/main/java/org/gridgain/internal/h2/table/TableFilter.java

```diff
@@ -312,21 +312,60 @@ public void prepare() {
         if (index.getClass() == HashJoinIndex.class)
             ((HashJoinIndex)index).prepare(session, indexConditions);
 
+        boolean hashIndex = index.getIndexType().isHash();
+        boolean[] columnPresenceMask = null;
```

Comment:
java.util.BitSet can be used instead.

Reply:
Seems, you need only BitSet here to mark columns, which a condition was found for.
Then you can use `index.getColumns()` to find the first column, which condition is missed.
Then index conditions for the found columns and all the next ones can be removed.

Reply:
AFAIK BitSet is good for large sets. In this case, we won't save much memory, and as far as I know, performance is worse with bitset than with boolean array.
But if you insist I can change the type to BitSet.